src/Accelerators/NNPA/Runtime/zDNNExtension/zDNNExtension.c

```diff
@@ -0,0 +1,280 @@
+  if (s)
+    cs = atoi(s);
+  assert(cs % AIU_STICKS_PER_PAGE == 0 && "Chunk size is not multiple of 32");
+  if (ZTensorSplitDebugFromEnv())
+    printf("OM_ZTENSOR_SPLIT_SIZE: %d\n", cs);
+  return cs;
+}
+
+bool ZTensorSplitEnabledFromEnv() {
+  int enabled = DEFAULT_ZTENSOR_SPLIT_ENABLED;
+  const char *s = getenv("OM_ZTENSOR_SPLIT_ENABLED");
+  if (s)
+    enabled = atoi(s);
+  if (ZTensorSplitDebugFromEnv())
+    printf("OM_ZTENSOR_SPLIT_ENABLED: %d\n", enabled);
+  return (enabled != 0);
+}
+
+bool ZTensorSplitDebugFromEnv() {
+  int enabled = DEFAULT_ZTENSOR_SPLIT_DEBUG;
+  const char *s = getenv("OM_ZTENSOR_SPLIT_DEBUG");
+  if (s)
+    enabled = atoi(s);
+  return (enabled != 0);
+}
+
+void getZTensorShape(const zdnn_ztensor *t, zTensorShape *shape) {
+  // Stickification: (e4, e3, e2, e1) -> (e4, e1/64, e3, e2/32, 32, 64)
+  const zdnn_tensor_desc *desc = t->transformed_desc;
+  shape->dim6 = desc->dim4;
+  shape->dim5 = CEIL(desc->dim1, AIU_2BYTE_CELLS_PER_STICK);
+  shape->dim4 = desc->dim3;
+  shape->dim3 = CEIL(desc->dim2, AIU_STICKS_PER_PAGE);
+  shape->dim2 = AIU_STICKS_PER_PAGE;
+  shape->dim1 = AIU_2BYTE_CELLS_PER_STICK;
+  // Tensor size is large, use uint64_t.
+  uint64_t sizeFromDim = (uint64_t)shape->dim6 * (uint64_t)shape->dim5 *
+                         (uint64_t)shape->dim4 * (uint64_t)shape->dim3 *
+                         (uint64_t)shape->dim2 * (uint64_t)shape->dim1 *
+                         (uint64_t)AIU_2BYTE_CELL_SIZE;
+  uint64_t sizeFromBuffer = t->buffer_size;
+  assert(sizeFromDim == sizeFromBuffer && "buffer size mismatched");
+}
+
+zdnn_status allocZTensorChunk(const zdnn_ztensor *input, uint32_t axis,
+    uint32_t chunkSize, zdnn_ztensor *output) {
+  zdnn_status status;
+  zdnn_tensor_desc *preTransDesc = malloc(sizeof(zdnn_tensor_desc));
+  zdnn_tensor_desc *transDesc = malloc(sizeof(zdnn_tensor_desc));
+  // Copy pre_transform_desc from the input.
+  preTransDesc->layout = input->pre_transformed_desc->layout;
+  preTransDesc->format = input->pre_transformed_desc->format;
+  preTransDesc->type = input->pre_transformed_desc->type;
+  preTransDesc->dim4 =
+      (axis == 0) ? chunkSize : input->pre_transformed_desc->dim4;
+  preTransDesc->dim3 =
+      (axis == 1) ? chunkSize : input->pre_transformed_desc->dim3;
+  preTransDesc->dim2 =
+      (axis == 2) ? chunkSize : input->pre_transformed_desc->dim2;
+  preTransDesc->dim1 =
+      (axis == 3) ? chunkSize : input->pre_transformed_desc->dim1;
+  // Copy a transformed desc.
+  status = zdnn_generate_transformed_desc(preTransDesc, transDesc);
+  assert(status == ZDNN_OK);
+  // Init a zTensor with malloc.
+  status = zdnn_init_ztensor_with_malloc(preTransDesc, transDesc, output);
+  return status;
+}
+
+zdnn_status freeZTensorChunk(zdnn_ztensor *t) {
+  if (t->pre_transformed_desc)
+    free(t->pre_transformed_desc);
+  if (t->transformed_desc)
+    free(t->transformed_desc);
+  zdnn_free_ztensor_buffer(t);
```

Comment:
Is it true that the `zdnn_free_ztensor_buffer` does not look at the `t->pre_transformed_desc` or `t->transformed_desc`?

If not sure, it might be safer to first free the buffer and then the descriptors?

Reply:
> Is it true that the zdnn_free_ztensor_buffer does not look at the t->pre_transformed_desc or t->transformed_desc?

Yes, it's true. But I still moved `free the buffer` up as your suggestion.